reset.c

```diff
@@ -0,0 +1,849 @@
+/*  reset.c --  removes aligner updates and reference data from input sam /
+                bam / cram file and makes read data raw for new processing
+
+    Copyright (C) 2022 Genome Research Ltd.
+
+    Author: Vasudeva Sarma <vasudeva.sarma@sanger.ac.uk>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE
+
+*/
+
+#include "samtools.h"
+#include "htslib/sam.h"
+#include "sam_opts.h"
+#include "htslib/thread_pool.h"
+#include "htslib/khash.h"
+
+KHASH_SET_INIT_INT(aux_set);        //set hold the aux tags to be retained
+KHASH_SET_INIT_STR(pg_set);         //set hold the pg tags to be removed
+
+#define TAGNUM(X) (((X)[0] << 8) | (X)[1])  //to create key for aux tags, like type key in htslib
+
+typedef struct conf_data
+{
+    int keepRGs;                    //RG line handling
+    int keepPGs;                    //PG line handling
+    khash_t(aux_set) *auxtagsin;    //SET that holds the aux tags to be retained
+    khash_t(pg_set) *pgout;         //SET that holds the pg tags to be removed
+
+    int tmpfield;                   //tmp field to hold any invocation specific data for the callback
+} conf_data;
+
+/// usage - print the usage
+/** @param fp pointer to the file / terminal to which usage to be dumped
+returns nothing
+*/
+static void usage(FILE *fp)
+{
+    fprintf(fp,
+            "Usage: samtools reset [options]\n\
+      -i <file path>        input file\n\
+      -o <file path>        output file\n\
+      -k <file path>        aux tags to be retained in csv file, optional\n\
```

Comment:
For sake of consistency, please see other subcommands for the description / style.

Generally we use `FILE` over `<file path>`.

We also usually have a `--no-PG` option to avoid adding the `@PG` line, and support `-@` as a short-hand form of `--threads`.

Reply:
Done.

Reply:
The `--no-PG` option is serving a different purpose here, but it's a useful one that I didn't think about.  Your option is removing the existing PG lines, in a similar fashion to how `--no-RG` removes the RG lines.  That's a good option to have.

However the other tools use `--no-PG` as a way to avoid *adding* a new PG record.  I'm not sure what the best option naming would be here as there's a clash of functionality.  However minimally when specifying `--no-PG` we shouldn't be added a new PG record I think.  Ideally we'd have two options for control the two different types of PG editing here.

Reply:
Ah I see it was originally named `--reject-PG`.  When I asked for `--no-PG` I was referring to the same functionality offered in other subcommands: to not add a new PG record.  So renaming it back to the original name is probably good.

I don't see the purpose however of selecting which individual PG records to keep in a CSV.  It feels complex.  I could perhaps see a specific single PG name which we discard and all subsequent in the chain.  Eg if our headers are:

```
@PG	ID:SCS	PN:HiSeq Control Software	...
@PG	ID:basecalling	PN:RTA	PP:SCS	...
@PG	ID:Illumina2bam	PN:Illumina2bam	PP:basecalling	...
@PG	ID:BamAdapterFinder	PN:BamAdapterFinder	PP:Illumina2bam	...
@PG	ID:BamIndexDecoder	PN:BamIndexDecoder	PP:BamAdapterFinder	...
@PG	ID:pb_cal	PN:predictor_pu	PP:BamIndexDecoder	...
@PG	ID:spf	PN:spatial_filter	PP:pb_cal	...
@PG	ID:bwa	PN:bwa	PP:spf	...
@PG	ID:BamMerger	PN:BamMerger	PP:bwa	...
@PG	ID:SplitBamByReadGroup	PN:SplitBamByReadGroup	PP:BamMerger	...
@PG	ID:bwa_aln	PN:bwa	PP:SplitBamByReadGroup	...
@PG	ID:bwa_aln_1	PN:bwa	PP:bwa_aln	...
@PG	ID:bwa_sam	PN:bwa	PP:bwa_aln_1	...
@PG	ID:Picard_SamFormatConverter	PN:SamFormatConverter	PP:bwa_sam	...
@PG	ID:samtools_fixmate	PN:samtools	PP:Picard_SamFormatConverter	...
@PG	ID:BamMerger_1	PN:BamMerger	PP:samtools_fixmate	...
@PG	ID:AlignmentFilter	PN:AlignmentFilter	PP:BamMerger_1	...
@PG	ID:samtools_sort	PN:samtools	PP:AlignmentFilter	...
@PG	ID:picard_markduplicates	PN:Picard	PP:samtools_sort	...
@PG	ID:ChangeBamHeader	PN:ChangeBamHeader	PP:picard_markduplicates	...
@PG	ID:BamTagStripper	PN:BamTagStripper	PP:ChangeBamHeader	...
@PG	ID:samtools	PN:samtools	PP:BamTagStripper	...
```

Then potentially we'd say `--reject-PG bwa` and everything from that point onwards get thrown away, while retaining the earlier processing that lead to the FASTQ file we input to bwa.

Reply:
I should also note that man page states it's using the PN field to match on.  The code looks to match both ID and PN which is perhaps OK, but I note PN is not unique so it would be unclear what "bwa" meant in my above example (it'd match ID:bwa, ID_bwa_aln, ID:bwa_aln_1 and ID:bwa_sam).  We could just say whatever is first I guess, as all descendents should get rejected anyway.

Reply:
Done; removes PGs from the 1st match to end on match of PN tag to given option (expects a name instead of earlier csv file).